File: Solutions/python/two-bucket/two_bucket.py
from collections import deque


def measure(bucket_one, bucket_two, goal, start_bucket):
    queue = deque([(0, 0, 0)])  # (amount in bucket 1, amount in bucket 2, step count)
    visited = set()

    while queue:
        amount_1, amount_2, steps = queue.popleft()

        if (amount_1, amount_2) in visited:
            continue

        # Skip invalid starting states
        if (start_bucket == "one" and amount_1 == 0 and amount_2 == bucket_two) or (
            start_bucket == "two" and amount_1 == bucket_one and amount_2 == 0
        ):
            continue

        visited.add((amount_1, amount_2))

        # Check if goal is reached
        if amount_1 == goal:
            return (steps, "one", amount_2)
        if amount_2 == goal:
            return (steps, "two", amount_1)

        # Generate all possible next states
        next_states = [
            (
                amount_1 - min(bucket_two - amount_2, amount_1),
                amount_2 + min(bucket_two - amount_2, amount_1),
            ),  # Pour 1 to 2
            (0, amount_2),  # Empty 1
            (bucket_one, amount_2),  # Fill 1
            (
                amount_1 + min(bucket_one - amount_1, amount_2),
                amount_2 - min(bucket_one - amount_1, amount_2),
            ),  # Pour 2 to 1
            (amount_1, 0),  # Empty 2
            (amount_1, bucket_two),  # Fill 2
        ]

        for new_amount_1, new_amount_2 in next_states:
            queue.append((new_amount_1, new_amount_2, steps + 1))

    raise ValueError("Goal amount cannot be reached.")

File: solutions/elixir/two-bucket/lib/two_bucket.ex
defmodule TwoBucket do
  defstruct [:bucket_one, :bucket_two, :moves]
  @type t :: %TwoBucket{bucket_one: integer, bucket_two: integer, moves: integer}

  @doc """
  Find the quickest way to fill a bucket with some amount of water from two buckets of specific sizes.
  """
  @spec measure(
          size_one :: integer,
          size_two :: integer,
          goal :: integer,
          start_bucket :: :one | :two
        ) :: {:ok, TwoBucket.t()} | {:error, :impossible}

  def measure(b1, b2, goal, :one), do: explore(b1, b2, goal, [{b1, 0, 1}], MapSet.new([{0, b2}]))
  def measure(b1, b2, goal, :two), do: explore(b1, b2, goal, [{0, b2, 1}], MapSet.new([{b1, 0}]))

  defp explore(_s1, _s2, _goal, [], _visited), do: {:error, :impossible}

  defp explore(s1, s2, goal, [{b1, b2, steps} | queue], visited) do
    cond do
      {b1, b2} in visited -> explore(s1, s2, goal, queue, visited)
      true -> explore_state(s1, s2, goal, b1, b2, steps, queue, visited)
    end
  end

  defp explore_state(_s1, _s2, goal, b1, b2, steps, _queue, _visited)
       when b1 == goal or b2 == goal do
    {:ok, %TwoBucket{bucket_one: b1, bucket_two: b2, moves: steps}}
  end

  defp explore_state(s1, s2, goal, b1, b2, steps, queue, visited) do
    new_states = next_states(s1, s2, b1, b2, steps)
    explore(s1, s2, goal, queue ++ new_states, MapSet.put(visited, {b1, b2}))
  end

  defp next_states(s1, s2, b1, b2, steps) do
    [
      # Fill bucket one
      {s1, b2, steps + 1},
      # Fill bucket two
      {b1, s2, steps + 1},
      # Pour from bucket two to bucket one
      {min(b1 + b2, s1), b2 - min(b1 + b2, s1) + b1, steps + 1},
      # Pour from bucket one to bucket two
      {b1 - min(b1 + b2, s2) + b2, min(b1 + b2, s2), steps + 1},
      # Empty bucket one
      {0, b2, steps + 1},
      # Empty bucket two
      {b1, 0, steps + 1}
    ]
  end
end
